File: src/main/java/me/mattstudios/triumphpets/TriumphPets.java
package me.mattstudios.triumphpets;

import me.mattstudios.mattscore.MattPlugin;
import me.mattstudios.mattscore.locale.Locales;
import me.mattstudios.triumphpets.commands.CMDGive;
import me.mattstudios.triumphpets.commands.CMDGives;
import me.mattstudios.triumphpets.commands.CMDPet;
import me.mattstudios.triumphpets.commands.CommandTest;
import me.mattstudios.triumphpets.commands.ListCommand;
import me.mattstudios.triumphpets.commands.PetsCommand;
import me.mattstudios.triumphpets.configuration.MainConfig;
import me.mattstudios.triumphpets.configuration.MessageConfig;
import me.mattstudios.triumphpets.data.PetDataHandler;
import me.mattstudios.triumphpets.gui.GuiHandler;
import me.mattstudios.triumphpets.listeners.PetListener;
import me.mattstudios.triumphpets.pet.PetController;
import me.mattstudios.triumphpets.pet.nms.v1_14_r1.EntityController_1_14_R1;
import me.mattstudios.triumphpets.pet.nms.v1_14_r1.PetRegistry_1_14_R1;

import java.util.stream.Stream;

import static me.mattstudios.mattscore.utils.MessageUtils.info;
import static me.mattstudios.mattscore.utils.NmsUtils.getServerVersion;

public final class TriumphPets extends MattPlugin {

    private PetController petController;
    private PetDataHandler petDataHandler;
    private GuiHandler guiHandler;

    @Override
    public void onPluginLoad() {
        registerNmsPets();
    }

    @Override
    public void onPluginEnable() {

        getConfig().load(MainConfig.class);
        getLocaleHandler().setLocale(Locales.EN_EN);
        getLocale().load(MessageConfig.class, Locales.EN_EN);
        System.out.println(getLocale().get(MessageConfig.STORAGE_METHOD));

        registerCommands();

        registerListeners();

        setUpNms();

        petDataHandler = new PetDataHandler(this);
        guiHandler = new GuiHandler(this);


    }

    @Override
    public void onPluginDisable() {
        petController.removeAll();
    }

    /**
     * Gets the interface to control all the pets.
     *
     * @return The pet controller.
     */
    public PetController getPetController() {
        return petController;
    }

    public PetDataHandler getPetDataHandler() {
        return petDataHandler;
    }

    public GuiHandler getGuiHandler() {
        return guiHandler;
    }

    /**
     * Registers all commands used in the plugin.
     */
    private void registerCommands() {
        Stream.of(
                new PetsCommand(this),
                new ListCommand(this),
                new CMDPet(this),
                new CMDGive(this),
                new CMDGives(this),
                new CommandTest()
        ).forEach(this::registerCommand);
    }

    /**
     * Registers all event listeners used by the plugin.
     */
    private void registerListeners() {
        Stream.of(
                new PetListener(this)
        ).forEach(this::registerListener);
    }

    /**
     * Registers pet's based on the NMS version
     */
    private void registerNmsPets() {
        String version = getServerVersion();
        switch (version) {
            case "v1_12_R2":
                System.out.println("soon 1.12");
                break;

            case "v1_13_R1":
                System.out.println("soon 1.13");
                break;

            default:
                if (!version.equalsIgnoreCase("v1_14_R1")) info("Might not support.");
                PetRegistry_1_14_R1.registerEntities();
                break;
        }
    }

    /**
     * Sets up the NMS classes, registering the pets and getting pet controller.
     */
    private void setUpNms() {
        String version = getServerVersion();
        switch (version) {
            case "v1_12_R2":
                System.out.println("soon 1.12");
                break;

            case "v1_13_R1":
                System.out.println("soon 1.13");
                break;

            default:
                if (!version.equalsIgnoreCase("v1_14_R1")) info("Might not support.");
                petController = new EntityController_1_14_R1(this);
                break;
        }
    }

}

File: src/main/kotlin/me/mattstudios/triumphpets/TriumphPets.kt
package me.mattstudios.triumphpets

import com.google.common.primitives.Ints
import me.mattstudios.mattcore.MattPlugin
import me.mattstudios.mattcore.utils.MessageUtils.color
import me.mattstudios.mattcore.utils.MessageUtils.info
import me.mattstudios.mattcore.utils.NmsUtils.getServerVersion
import me.mattstudios.mf.base.components.CompletionResolver
import me.mattstudios.mf.base.components.ParameterResolver
import me.mattstudios.mf.base.components.TypeResult
import me.mattstudios.triumphpets.commands.admin.CrateCommand
import me.mattstudios.triumphpets.commands.admin.PetGiveCommand
import me.mattstudios.triumphpets.commands.player.PetsCommand
import me.mattstudios.triumphpets.config.Settings
import me.mattstudios.triumphpets.config.pet.PetConfig
import me.mattstudios.triumphpets.config.pet.PetDefaultConfig
import me.mattstudios.triumphpets.config.pet.PetFileConfig
import me.mattstudios.triumphpets.listeners.CrateListeners
import me.mattstudios.triumphpets.listeners.PetListeners
import me.mattstudios.triumphpets.listeners.PlayerListeners
import me.mattstudios.triumphpets.locale.Message
import me.mattstudios.triumphpets.managers.PetManager
import me.mattstudios.triumphpets.pet.utils.Experience
import me.mattstudios.triumphpets.pet.utils.PetType
import me.mattstudios.triumphpets.pet.utils.PetUtils
import me.mattstudios.triumphpets.util.Utils.getPluginStartup
import java.util.regex.Pattern
import java.util.stream.Stream


/**
 * @author Matt
 */
class TriumphPets : MattPlugin() {

    lateinit var petManager: PetManager
        private set
    lateinit var petConfig: PetConfig
        private set

    /**
     * On plugin enable
     */
    override fun enable() {
        setupConfig()
        setupLocale()

        startUpMessage()

        Experience.load(config)

        petManager = PetManager(this)

        setupCommands()
        setupListeners()
    }

    /**
     * On plugin disable
     */
    override fun disable() {
        petManager.disable()
    }

    /**
     * Sets up the config related stuff
     */
    private fun setupConfig() {
        config.load(Settings::class.java)

        petConfig = if (config[Settings.USE_PET_CONFIG]) PetFileConfig(this) else PetDefaultConfig()
    }

    /**
     * Sets up the locale of the plugin
     */
    private fun setupLocale() {
        PetUtils.LOCALE = config[Settings.LANGUAGE]
        locale.load(Message::class.java, PetUtils.LOCALE)
    }

    /**
     * Sets up everything related to commands
     */
    private fun setupCommands() {
        val pattern = Pattern.compile("_(\\d+)_")
        val matcher = pattern.matcher(getServerVersion())

        var version: Int? = null

        while (matcher.find()) {
            @Suppress("UnstableApiUsage")
            version = Ints.tryParse(matcher.group(1))
        }

        commandManager.parameterHandler.register(PetType::class.java, ParameterResolver { argument ->
            if (version == null || argument == null) return@ParameterResolver TypeResult(argument)
            val petType = PetType.values().toList().filter { it.version <= version }.find { it.name.equals(argument.toString(), true) }
            return@ParameterResolver TypeResult(petType, argument)
        })

        registerCompletion("#pets", CompletionResolver {
            if (version == null) return@CompletionResolver listOf("")
            return@CompletionResolver PetType.values().toList().filter { it.version <= version }.map { it.name }
        })

        registerCompletion("#crate-type") { listOf("set", "edit", "unset") }

        registerCommands(
                PetGiveCommand(this),
                CrateCommand(this),

                PetsCommand(this)
                        )
    }

    /**
     * Registers all the listeners
     */
    private fun setupListeners() {
        registerListeners(
                PetListeners(this),
                PlayerListeners(this),
                CrateListeners(this)
                         )
    }

    /**
     * Just the startup message
     */
    private fun startUpMessage() {
        Stream.of(
                "",
                getPluginStartup(),
                "${locale.getMessageRaw(Message.STARTUP_VERSION)} &c${description.version} &8(&c${PetUtils.LOCALE}&8)"
                 ).forEach { info(color(it)) }
    }

}